File: examples/backends/fastapi/app/session/dynamodb.py
"""DynamoDB session backend for production deployments."""

from __future__ import annotations

import json
import time
from typing import Any

import aioboto3


class DynamoDBSessionBackend:
    """Session backend using AWS DynamoDB.

    Table schema:
        Partition key: session_id (S)
        Attributes: data (S, JSON-encoded), created_at (N), ttl (N)

    Enable TTL on the `ttl` attribute for automatic cleanup.
    """

    def __init__(
        self,
        table_name: str = "l42_sessions",
        max_age: int = 30 * 24 * 3600,
        endpoint_url: str = "",
        region_name: str = "us-west-2",
    ) -> None:
        self._table_name = table_name
        self._max_age = max_age
        self._session = aioboto3.Session()
        self._endpoint_url = endpoint_url or None
        self._region_name = region_name

    async def load(self, session_id: str) -> dict[str, Any] | None:
        async with self._session.resource(
            "dynamodb",
            endpoint_url=self._endpoint_url,
            region_name=self._region_name,
        ) as dynamodb:
            table = await dynamodb.Table(self._table_name)
            response = await table.get_item(Key={"session_id": session_id})

        item = response.get("Item")
        if item is None:
            return None

        created_at = float(item.get("created_at", 0))
        if time.time() - created_at > self._max_age:
            await self.delete(session_id)
            return None

        return json.loads(item["data"])

    async def save(self, session_id: str, data: dict[str, Any]) -> None:
        now = time.time()
        async with self._session.resource(
            "dynamodb",
            endpoint_url=self._endpoint_url,
            region_name=self._region_name,
        ) as dynamodb:
            table = await dynamodb.Table(self._table_name)
            await table.put_item(
                Item={
                    "session_id": session_id,
                    "data": json.dumps(data),
                    "created_at": int(now),
                    "ttl": int(now + self._max_age),
                }
            )

    async def delete(self, session_id: str) -> None:
        async with self._session.resource(
            "dynamodb",
            endpoint_url=self._endpoint_url,
            region_name=self._region_name,
        ) as dynamodb:
            table = await dynamodb.Table(self._table_name)
            await table.delete_item(Key={"session_id": session_id})

File: rust/src/session/dynamodb.rs
//! DynamoDB session backend for production/Lambda deployments.
//!
//! Same table schema as the FastAPI backend:
//! - `session_id` (S) — partition key
//! - `data` (S) — JSON-encoded session payload
//! - `created_at` (N) — Unix timestamp when session was created
//! - `ttl` (N) — Unix timestamp for DynamoDB automatic cleanup
//!
//! Compatible with existing FastAPI DynamoDB sessions table.

use aws_sdk_dynamodb::types::AttributeValue;
use aws_sdk_dynamodb::Client;
use std::time::{SystemTime, UNIX_EPOCH};

use super::{SessionBackend, SessionData};

const MAX_AGE_SECS: u64 = 30 * 24 * 3600; // 30 days

/// DynamoDB session backend.
pub struct DynamoDbBackend {
    client: Client,
    table_name: String,
}

impl DynamoDbBackend {
    pub fn new(client: Client, table_name: String) -> Self {
        Self { client, table_name }
    }
}

fn now_secs() -> u64 {
    SystemTime::now()
        .duration_since(UNIX_EPOCH)
        .unwrap_or_default()
        .as_secs()
}

impl SessionBackend for DynamoDbBackend {
    async fn load(&self, session_id: &str) -> Option<SessionData> {
        let result = self
            .client
            .get_item()
            .table_name(&self.table_name)
            .key("session_id", AttributeValue::S(session_id.to_string()))
            .send()
            .await
            .ok()?;

        let item = result.item()?;

        // Check application-level TTL (in case DynamoDB hasn't cleaned up yet)
        let created_at = item
            .get("created_at")
            .and_then(|v| v.as_n().ok())
            .and_then(|n| n.parse::<u64>().ok())
            .unwrap_or(0);

        let now = now_secs();
        if now.saturating_sub(created_at) > MAX_AGE_SECS {
            // Expired — delete and return None
            let _ = self
                .client
                .delete_item()
                .table_name(&self.table_name)
                .key("session_id", AttributeValue::S(session_id.to_string()))
                .send()
                .await;
            return None;
        }

        // Deserialize session data from JSON string
        let data_json = item.get("data")?.as_s().ok()?;
        serde_json::from_str(data_json).ok()
    }

    async fn save(&self, session_id: &str, data: &SessionData) {
        let now = now_secs();
        let ttl = now + MAX_AGE_SECS;
        let data_json = match serde_json::to_string(data) {
            Ok(j) => j,
            Err(e) => {
                tracing::error!("Failed to serialize session data: {}", e);
                return;
            }
        };

        // Use UpdateItem with if_not_exists to preserve original created_at.
        // PutItem would overwrite created_at on every save, making the 30-day
        // TTL reset on each write (effectively "30 days since last activity"
        // instead of "30 days since creation").
        let _ = self
            .client
            .update_item()
            .table_name(&self.table_name)
            .key("session_id", AttributeValue::S(session_id.to_string()))
            .update_expression(
                "SET #data = :data, created_at = if_not_exists(created_at, :now), #ttl = :ttl",
            )
            .expression_attribute_names("#data", "data")
            .expression_attribute_names("#ttl", "ttl")
            .expression_attribute_values(":data", AttributeValue::S(data_json))
            .expression_attribute_values(":now", AttributeValue::N(now.to_string()))
            .expression_attribute_values(":ttl", AttributeValue::N(ttl.to_string()))
            .send()
            .await
            .map_err(|e| {
                tracing::error!("Failed to save session {}: {}", session_id, e);
            });
    }

    async fn delete(&self, session_id: &str) {
        let _ = self
            .client
            .delete_item()
            .table_name(&self.table_name)
            .key("session_id", AttributeValue::S(session_id.to_string()))
            .send()
            .await
            .map_err(|e| {
                tracing::error!("Failed to delete session {}: {}", session_id, e);
            });
    }
}
